easybuild/tools/job/pbs_python.py

```diff
@@ -45,69 +47,122 @@
-    node_vals = pq.getnodes().values()  # only the values, not the names
-    interesting_nodes = ('free', 'job-exclusive',)
-    res = {}
-    for np in [int(x['np'][0]) for x in node_vals if x['state'][0] in interesting_nodes]:
-        res.setdefault(np, 0)
-        res[np] += 1
-
-    # return most frequent
-    freq_count, freq_np = max([(j, i) for i, j in res.items()])
-    log.debug("Found most frequent np %s (%s times) in interesting nodes %s" % (freq_np, freq_count, interesting_nodes))
-
-    return freq_np
+    _log.debug("Failed to import pbs from pbs_python."
+               " Silently ignoring, this is a real issue only with --job=pbs")
+    # no `pbs_python` available, turn function into a no-op
+    def only_if_pbs_import_successful(fn):
+        def instead(*args, **kwargs):
+            """This is a no-op since `pbs_python` is not available."""
+            errmsg = ("PBSQuery or pbs modules not available."
+                      " Please make sure `pbs_python` is installed and usable.")
+            _log.error(errmsg)
+            raise RuntimeError(errmsg)
+        return instead
+    HAVE_PBS_PYTHON = False
+
+
+class Pbs(JobServer):
+    """
+    Manage PBS server communication and create `PbsJob` objects.
+    """
+
+    USABLE = HAVE_PBS_PYTHON
+
+    def __init__(self, pbs_server=None):
+        self.pbs_server = pbs_server or pbs.pbs_default()
+        self.conn = None
+        self._ppn = None
+
+    def begin(self):
+        self.connect_to_server()
+        self._submitted = []
+
+    @only_if_pbs_import_successful
+    def connect_to_server(self):
+        """Connect to PBS server, set and return connection."""
+        if not self.conn:
+            self.conn = pbs.pbs_connect(self.pbs_server)
+        return self.conn
+
+    def submit(self, job, after=frozenset()):
+        assert isinstance(job, PbsJob)
+        if after:
+            job.add_dependencies(after)
+        job._submit()
+        self._submitted.append(job)
+
+    def commit(self):
+        # release all user holds on jobs after submission is completed
+        for job in self._submitted:
+            if job.has_holds():
+                _log.info("releasing user hold on job %s" % job.jobid)
+                job.release_hold()
+        self.disconnect_from_server()
+        if self._submitted:
+            _log.info("List of submitted jobs:")
+            for job in self._submitted:
+                _log.info("* %s (%s): %s", job.name, job.module, job.jobid)
+
+    @only_if_pbs_import_successful
+    def disconnect_from_server(self):
+        """Disconnect current connection."""
+        pbs.pbs_disconnect(self.conn)
+        self.conn = None
+
+    def _get_ppn(self):
+        """Guess PBS' `ppn` value for a full node."""
+        # cache this value as it's not likely going to change over the
+        # `eb` script runtime ...
+        if not self._ppn:
+            log = fancylogger.getLogger('pbs_job.PbsServer.ppn')
+
+            pq = PBSQuery()
```

Comment:
if you want to clean up things in `pbs_python`, please handle that in a separate PR (after this one is merged in)

I don't want to get this PR blocked because of changes to something we don't plan to use anymore...